### ISSUE TITLE
Too many open files

### ISSUE THREAD

**Issue by appliedprivacy** [facebookexperimental/doh-proxy#66]
In the light of the errors bellow we were wondering what amount of open files is doh-httpproxy expected to require.

The logs bellow already generated 50GB of logs at one point.

```
 Traceback (most recent call last):
   File "/usr/local/lib/python3.7/dist-packages/aiohttp/web_protocol.py", line 418, in start
     resp = await task
   File "/usr/local/lib/python3.7/dist-packages/aiohttp/web_app.py", line 458, in _handle
     resp = await handler(request)
   File "/usr/local/lib/python3.7/dist-packages/aiohttp/web_middlewares.py", line 119, in impl
     return await handler(request)
   File "/usr/local/lib/python3.7/dist-packages/aiohttp_remotes/x_forwarded.py", line 111, in middleware
     return await handler(request)
   File "/usr/local/lib/python3.7/dist-packages/dohproxy/httpproxy.py", line 68, in doh1handler
     return await request.app.resolve(request, dnsq)
   File "/usr/local/lib/python3.7/dist-packages/dohproxy/httpproxy.py", line 81, in resolve
     dnsr = await dnsclient.query(dnsq, clientip)
   File "/usr/local/lib/python3.7/dist-packages/dohproxy/server_protocol.py", line 45, in query
     dnsr = await self.query_udp(dnsq, clientip, timeout=timeout)
   File "/usr/local/lib/python3.7/dist-packages/dohproxy/server_protocol.py", line 56, in query_udp
     remote_addr=(self.upstream_resolver, self.upstream_port))
   File "/usr/lib/python3.7/asyncio/base_events.py", line 1245, in create_datagram_endpoint
     raise exceptions[0]
   File "/usr/lib/python3.7/asyncio/base_events.py", line 1218, in create_datagram_endpoint
     family=family, type=socket.SOCK_DGRAM, proto=proto)
   File "/usr/lib/python3.7/socket.py", line 151, in __init__
     _socket.socket.__init__(self, family, type, proto, fileno)
 OSError: [Errno 24] Too many open files
```

```
ERROR: socket.accept() out of system resource
 socket: <socket.socket fd=6, family=AddressFamily.AF_INET6, type=SocketKind.SOCK_STREAM, proto=6, laddr=(::1, 8084, 0, 0)>
 Traceback (most recent call last):
   File "/usr/lib/python3.7/asyncio/selector_events.py", line 156, in _accept_connection
   File "/usr/lib/python3.7/socket.py", line 212, in accept
     fd, addr = self._accept()
 OSError: [Errno 24] Too many open files
```

**Comment by chantra** [comment]
Hi @appliedprivacy 

I don't have a good answer for you as this is very much dependent on your workload.

This tells that there doh-proxy needs more file descriptor, basically it reaches it limit. What is you limit currently set at? Do you have monitoring that would show whether or not the number of fd in use slowly grows up overtime?

@appliedprivacy  I would also like to set the expectation right here. Based on the README:

> DOH Proxy was created during IETF Hackathon 100 as a proof-of-concept and is not used at Facebook.
> You are welcome to use it, but be aware that support is limited and best-effort.

If you are running this in a production environment, you will most likely surface issues of this type and will be in the best position to reproduce and fix them.

Looking at a small instance of doh-proxy that I am running, I see indeed a whole bunch of hanging UDP socket:
```
udp    ESTAB      0      0       ::1:41782               ::1:53                  users:(("doh-httpproxy",pid=857,fd=242))
```

This would seem to indicate that there may be conditions when the code is not closing the socket upon reception of reply, timeout, exception.... I would guess it is most likely possible that when we timeout a request or hit any type of exception, we do not close the socket.

Some potential places:
 handling of exception when writing the message back
```
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: 2019-05-25 04:58:22,789:    ERROR: Exception in callback _SelectorDatagramTransport._read_ready()
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: handle: <Handle _SelectorDatagramTransport._read_ready()>
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: Traceback (most recent call last):
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: File "/usr/lib64/python3.6/asyncio/events.py", line 145, in _run
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: self._callback(*self._args)
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: File "/usr/lib64/python3.6/asyncio/selector_events.py", line 1077, in _read_ready
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: self._protocol.datagram_received(data, addr)
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: File "/usr/lib/python3.6/site-packages/dohproxy/server_protocol.py", line 139, in datagram_received
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: self.receive_helper(dnsr)
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: File "/usr/lib/python3.6/site-packages/dohproxy/server_protocol.py", line 128, in receive_helper
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: self.fut.set_result(dnsr)
May 25 04:58:22 centos-512mb-sfo1-01 doh-httpproxy[853]: asyncio.base_futures.InvalidStateError: invalid state
```

Another possible one maybe: https://github.com/facebookexperimental/doh-proxy/blob/master/dohproxy/server_protocol.py#L147

I would recommend you attempt to fix the code to better handle exceptions there and close the transport when needed, test it and see if this fixes your issue.

**Comment by appliedprivacy** [comment]
thanks for your reply.

> I would also like to set the expectation right here. Based on the README:

No worries we are not in the position to expect anything. We are just filing issues as we assume this helps and contributes.

**Comment by chantra** [comment]
On Sun, May 26, 2019 at 5:48 AM appliedprivacy <notifications@github.com>
wrote:

> thanks for your reply.
>
> I would also like to set the expectation right here. Based on the README:
>
> No worries we are not in the position to expect anything. We are just
> filing issues as we assume this helps and contributes.
>
>
Great! Thanks for reporting indeed :)

**Comment by chantra** [comment]
@appliedprivacy this should be fixed through @eraymitrani‘s fix in #67 . If you update to master, you should be good now.

**Comment by chantra** [comment]
v0.0.9 was pushed to pip
